Add interop framework (#26)
ethereum/L2-interop: docs/framework.md

# Interop Framework

This document tries to lay out a framework for understanding and discussing cross-chain interop protocols.

- **Data vs Value Transfer**
    - A data-transfer protocol allows information in general to be transmitted cross-chain.
    - A value-transfer protocol allows cross-chain transmission of fungible assets.

>         - Cross-chain value transfer is a form of exchange.

Author: Yeah I thought about this a little too... Obviously everything is data in the end, but I'm not sure that value transfers are specialized data transfers. They can be implemented as data transfer but they also admit different implementations. Overall I think it's useful to consider them as separate things.

- **Narrow vs Broad**

>     - In a narrow protocol, specific actions on a source chain initiate cross-chain communication. For example, that a specific contract was invoked (gateway, outbox, bridge, etc.) or a specific kind of event was emitted.This subset of the chain state is transmitted cross-chain.

Reviewer: ```suggestion
    - In a narrow protocol, specific actions on a source chain initiate cross-chain communication. For example, the invocation of a specific contract (gateway, outbox, bridge, etc.) or the emission of a specific type of event. This subset of the chain state is transmitted cross-chain.
```

        - Each of these actions is a **message**, so narrow protocols are also known as **message-based** protocols.
    - In a broad protocol, an unrestricted set of source chain state can be the basis for cross-chain communication. For example, all logs or all storage from the source chain could be accessed on the destination chain.
- **Push vs Pull**
    - A push protocol allows for the entire cross-chain transmission lifecycle, up to and including delivery to a receiver, to be arranged on the source chain without a user transaction on the destination chain. The protocol may rely on **third party relayers** to complete the lifecycle.

>     - A pull protocol requires a user (or application) to complete the lifecycle with a transaction on the destination chain. The protocol can be said to require **self-relaying**.

Author: IMO this needs to be defined from the perspective of the user. If a third-party relayer completes the operation it's indistinguishable from something like OP L1 deposits being relayed automatically, so it should count as "push".

Perhaps there's a distinction to be made between "system" and "third-party" push.

I also think the problem is that push vs pull is not an either-or property of a protocol, they're more like two modes that can coexist. A protocol can support both, and each cross-chain transmission through the protocol can use a different mode.

Does this make sense?

>     - A pull protocol requires a user (or application) to complete the lifecycle with a transaction on the destination chain. The protocol can be said to require **self-relaying**.

Reviewer 2: I've updated nomenclatures.md to reflect said definitions ^^
